### ISSUE TITLE
[d3d8] GameMaker games trip up fullscreen presentation on the GDI fallback path

### ISSUE THREAD

**Issue by WinterSnowfall** [doitsujin/dxvk#4246]
The game will fail to present when in fullscreen, thus the result will be a perpetual black screen. The logs spam a lot of:

```
err:   D3D9SwapChainEx::BlitGDI Surface GetDC failed
err:   D3D9SwapChainEx::BlitGDI Surface GetDC failed
```

interweaved with device resets, until the game eventually gives up. Windowed more works just fine.

WineD3D also works just fine, both windowed and fullscreen.

### Software information
[Hydorah (2010)](https://locomalito.com/files/hydorah.zip) , a freeware side-scrolling shooter.

### System information
- GPU: Intel(R) Xe Graphics (TGL GT2)
- Driver: Intel open-source Mesa driver 24.0.9
- Wine version: 9.17 Staging
- DXVK version: daccde7

### Apitrace file(s)
[Hydorah.trace.tar.xz](https://drive.proton.me/urls/RRSN07Q5A0#g3dNSDMytPSA) (captured with dxvk, while the problem was manifesting)

### Log files
[Hydorah.log](https://github.com/user-attachments/files/16921210/Hydorah.log)

**Comment by WinterSnowfall** [comment]
P.S.: Apparently it's a problem with GameMaker, the game's underlying middleware, which had a d3d8 renderer around ~2010 or so. I've tried a few other games built on GameMaker and they all exhibit the same problem when in fullscreen mode.

**Comment by dungeon007** [comment]
Game seems to be created for devices of 2:1 ratio screen, 640x320. On let say a 1920x1080 screen, that is like 1920x960.
 On WINED3D correct, on NINE fullscreen small 640x320, similar on dgVoodoo2 d3d11 or d3d12 also 640x320 but it gets to a window.
 DXVK's d3d8 going BIG. 
 But, It looks like working with d3d8to9 + DXVK d3d9 combination too. And it dont complain on exit when screen is 2:1 
 
![1920x960](https://github.com/user-attachments/assets/2650fa89-efc4-4459-b29f-3d18cf28677f)

Scaling or probably that feature that add blacks on the sides on non 2:1 screens somehow fails.

**Comment by WinterSnowfall** [comment]
Well, that would explain, I guess, why dxvk thinks partial presentation is involved and trips up the GDI path (render rect would be smaller than the screen resolution, I believe). Ideally we shouldn't end up here in the first place with these sort of games, but then again it's also odd it doesn't work as expected. Maybe @K0bin can have a look when time permits, as this entire GDI thing is his design.

**Comment by dungeon007** [comment]
It seems had not worked as expected in 8to9 1.8.0 too, probably before doing something like this:
https://github.com/crosire/d3d8to9/commit/007c3d2ed9062f10e327aef3cfe41ae5b71dac3e
Just a guess, to 1.9.0 there are couple patches diff... but on first sight this most looks like a thing.
Or you can always look what wined3d is doing there
https://github.com/wine-mirror/wine/blob/master/dlls/d3d8/device.c#L1780

**Comment by WinterSnowfall** [comment]
> It seems had not worked as expected in 8to9 1.8.0 too, probably before doing something like this

Yes, I noticed that. See: #4245. But the GDI problem is unrelated and takes precedence, so it doesn't even really get up to the SetViewport calls at the moment.

**Comment by dungeon007** [comment]
Since ddraw is around and that could possibly interfere, forcing  **WINE_D3D_CONFIG="renderer=gdi"** might help.
I really hate when i see that GLfied ddraw around in d3d8 apps 🤣

**Comment by WinterSnowfall** [comment]
Apparently it's fine on Nvidia in terms of rendering (viewport is indeed squewed, but #4245 should take care of that), so those GDI calls were only failing on ANV??? That's really odd...

**Comment by K0bin** [comment]
Fixed by #4258
